File: examples/ocaml/linear/dense.c
#include <stdio.h>
#include <stdlib.h>
#include <sundials/sundials_types.h>

#if 600 <= SUNDIALS_LIB_VERSION
#include <sunmatrix/sunmatrix_dense.h>
#else
#include <sundials/sundials_direct.h>
#include <sundials/sundials_dense.h>
#endif

#include "../../../src/config.h"
#include "../../../src/sundials/sundials_ml.h"

#define NROWS 3
#define NCOLS 3

sunrealtype a_init[NROWS][NCOLS] = {
    {  1.0,  2.0,  3.0},
    {  2.0, -4.0,  6.0},
    {  3.0, -9.0, -3.0}
};

#if 700 <= SUNDIALS_LIB_VERSION
void print_mat(SUNDlsMat m, sundials_ml_index nr, sundials_ml_index nc) {
#else
void print_mat(DlsMat m, sundials_ml_index nr, sundials_ml_index nc) {
#endif
    int i, j;

    for (i=0; i < nr; ++i) {
	for (j=0; j < nc; ++j) {
#if 700 <= SUNDIALS_LIB_VERSION
        printf(" % e", SUNDLS_DENSE_ELEM(m, i, j));
#else
        printf(" % e", DENSE_ELEM(m, i, j));
#endif
	}
	printf("\n");
    }
}

void print_vec(sunrealtype* m, sundials_ml_index nr) {
    int i;

    for (i=0; i < nr; ++i) {
	printf(" % e", m[i]);
    }
    printf("\n");
}

void print_pivots(sundials_ml_index* m, sundials_ml_index nr) {
    int i;

    for (i=0; i < nr; ++i) {
	printf(" % lld", (long long)m[i]);
    }
    printf("\n");
}

int main(int argc, char** argv)
{
#if 700 <= SUNDIALS_LIB_VERSION
    SUNDlsMat a = SUNDlsMat_NewDenseMat(NROWS, NCOLS);
    SUNDlsMat b = SUNDlsMat_NewDenseMat(NROWS, NCOLS);
#elif 600 <= SUNDIALS_LIB_VERSION
    DlsMat a = SUNDlsMat_NewDenseMat(NROWS, NCOLS);
    DlsMat b = SUNDlsMat_NewDenseMat(NROWS, NCOLS);
#else
    DlsMat a = NewDenseMat(NROWS, NCOLS);
    DlsMat b = NewDenseMat(NROWS, NCOLS);
#endif
    sundials_ml_index p[NROWS] = { 0.0 };
    sunrealtype s[NROWS] = { 5.0, 18.0, 6.0 };
    int i, j;

    for (i=0; i < NROWS; ++i) {
	for (j=0; j < NCOLS; ++j) {
#if 700 <= SUNDIALS_LIB_VERSION
        SUNDLS_DENSE_ELEM(a, i, j) = a_init[i][j];
#else
        DENSE_ELEM(a, i, j) = a_init[i][j];
#endif
	}
    }

    printf("initially: a=\n");
    print_mat(a, NROWS, NCOLS);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 260
    {
	sunrealtype x[NCOLS] = { 1.0,  2.0, 3.0 };
	sunrealtype y[NROWS] = { 0.0 };
	printf("matvec: y=\n");
#if 600 <= SUNDIALS_LIB_VERSION
	SUNDlsMat_DenseMatvec(a, x, y);
#else
	DenseMatvec(a, x, y);
#endif
	print_vec(y, NROWS);
	printf("\n");
    }
#endif

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_DenseCopy(a, b);
#else
    DenseCopy(a, b);
#endif

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_DenseScale(2.0, b);
#else
    DenseScale(2.0, b);
#endif
    printf("scale copy x2: b=\n");
    print_mat(b, NROWS, NCOLS);
    printf("\n");

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_denseAddIdentity(b->cols, NROWS);
#else
    denseAddIdentity(b->cols, NROWS);
#endif
    printf("add identity: b=\n");
    print_mat(b, NROWS, NCOLS);
    printf("\n");

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_DenseGETRF(a, p);
#else
    DenseGETRF(a, p);
#endif
    printf("getrf: a=\n");
    print_mat(a, NROWS, NCOLS);
    printf("\n       p=\n");
    print_pivots(p, NROWS);
    printf("\n");

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_DenseGETRS(a, p, s);
#else
    DenseGETRS(a, p, s);
#endif
    printf("getrs: s=\n");
    print_vec(s, NROWS);

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_DestroyMat(a);
    SUNDlsMat_DestroyMat(b);
#else
    DestroyMat(a);
    DestroyMat(b);
#endif

    return 0;
}

File: examples/ocaml/linear/dense.ml

open Sundials
open Bigarray

module M = Matrix.Dense
let printf = Format.printf
let fprintf = Format.fprintf

let print_mat out m =
  let (nr, nc) = M.size m in
  fprintf out "@[<v>";
  for i = 0 to nr - 1 do
    fprintf out "@[<h>";
    for j = 0 to nc - 1 do
      fprintf out "@ % e" (M.get m i j)
    done;
    fprintf out "@]@ "
  done;
  fprintf out "@]"

let print_vec out m =
  let nr = Array1.dim m in
  fprintf out "@[<h>";
  for i = 0 to nr - 1 do
    fprintf out "@ % e" m.{i}
  done;
  fprintf out "@]"

let print_p out p =
  let nc = Array1.dim p in
  fprintf out "@[<h>";
  for i = 0 to nc - 1 do
    fprintf out "@ % d" p.{i}
  done;
  fprintf out "@]"

let nrows, ncols = 3, 3;;

let main () =
  let a = M.create nrows ncols in
  let zero = M.make nrows ncols 0.0 in

  M.set a 0 0 ( 1.0);
  M.set a 0 1 ( 2.0);
  M.set a 0 2 ( 3.0);

  M.set a 1 0 ( 2.0);
  M.set a 1 1 (-4.0);
  M.set a 1 2 ( 6.0);

  M.set a 2 0 ( 3.0);
  M.set a 2 1 (-9.0);
  M.set a 2 2 (-3.0);

  printf "initially: a=@\n%a@\n" print_mat a;

  (try
    let x = RealArray.of_array [| 1.0; 2.0; 3.0 |] in
    let y = RealArray.create nrows in
    M.matvec a x y;
    printf "matvec: y=@\n%a@\n\n" print_vec y
  with Config.NotImplementedBySundialsVersion -> ());

  let b = M.create nrows ncols in
  M.blit ~src:a ~dst:b;

  M.scale_add 2.0 b zero;
  printf "scale copy x2: b=@\n%a@\n" print_mat b;

  M.scale_addi 1.0 b;
  printf "add identity: b=@\n%a@\n" print_mat b;

  let a_ra2 = Sundials.RealArray2.wrap (M.unwrap a) in

  let p = LintArray.create nrows in
  Array1.fill p 0;
  Matrix.ArrayDense.getrf a_ra2 p;
  printf "getrf: a=@\n%a@\n" print_mat a;
  printf "       p=@\n%a@\n@\n" print_p p;

  let s = RealArray.create nrows in
  s.{0} <-  5.0;
  s.{1} <- 18.0;
  s.{2} <-  6.0;
  Matrix.ArrayDense.getrs a_ra2 p s;
  printf "getrs: s=@\n%a@\n" print_vec s;
  ();;

main ();;
Gc.compact ();;

